cai_causal_graph/identify_utils.py

```diff
@@ -387,3 +387,45 @@ def identify_markov_boundary(graph: Union[CausalGraph, Skeleton], node: NodeLike
         mb = graph.get_neighbors(node_id)
 
     return mb
+
+
+def identify_colliders(graph: CausalGraph, unshielded: bool = False) -> List[str]:
+    """
+    Identify all the collider nodes in the provided `graph`.
+
+    :param graph: The graph given by a `cai_causal_graph.causal_graph.CausalGraph` instance.
+    :param unshielded: If `True`, only unshielded colliders are returned. If `False`, all colliders are returned.
+        Default is `False`.
+    """
+    colliders = set()
+
+    bidirected_edges = [(edge.source.identifier, edge.destination.identifier) for edge in graph.get_bidirected_edges()]
+
+    # go through every node in the graph
+    for node in graph.get_node_names():
+        # the node z is collider if x <> z <> y, x -> z <- y, x <> z <- y, or x -> z <> y
+        neighbors = graph.get_neighbors(node)
+
+        counter = 0
+        for neighbor_node in neighbors:
+            # we need to check if the directed edge (neigh_node, node) exists
+            directed_exists = (
+                graph.edge_exists(neighbor_node, node)
+                and graph.get_edge(neighbor_node, node).edge_type == EdgeType.DIRECTED_EDGE
+            )
+            # we need to check if the bi-directed edge (neigh_node, node) or (node, neigh_node) exists
+            bidirected_exists = (neighbor_node, node) in bidirected_edges or (node, neighbor_node) in bidirected_edges
+            is_unshielded = not graph.edge_exists(neighbor_node, node) and not graph.edge_exists(node, neighbor_node)
```

Comment:
Correct me if I am wrong - unshielded collider is a collider if the parents are themselves not adjacent. So, the simplest case is:
```
        cg = CausalGraph()
        cg.add_edge('x', 'y')
        cg.add_edge('z', 'y')
```

here - `y` is an unshielded collider, since `x` and `z` are not adjacent?

Currently, if I do `dentify_colliders(cg, unshielded=True)`, it will return an empty list, i.e. fail to identify this collider.

Or did I get the definition wrong?

I think the logic that is wrong is here - you check for adjacency between node and potential parent, rather than between parents themselves?

Reply:
Yes you are right, I inadvertently removed the right code.